fendermint/vm/topdown/src/convert.rs

```diff
@@ -0,0 +1,159 @@
+}
+
+fn convert_addr(addr: Address) -> gateway_router_facet::FvmAddress {
+    gateway_router_facet::FvmAddress {
+        addr_type: addr.protocol() as u8,
+        payload: addr_payload_to_bytes(addr.into_payload()),
+    }
+}
+
+pub fn encode_commit_parent_finality_call(
+    finality: IPCParentFinality,
+    validator_set: ValidatorSet,
+) -> anyhow::Result<Vec<u8>> {
+    let commit_function = gateway_router_facet::GATEWAYROUTERFACET_ABI
+        .functions
+        .get(COMMIT_PARENT_FINALITY_FUNC_NAME)
+        .ok_or_else(|| {
+            anyhow!(
+                "report bug, abi function map does not have {}",
+                COMMIT_PARENT_FINALITY_FUNC_NAME
+            )
+        })?
+        .get(0)
+        .ok_or_else(|| {
+            anyhow!(
+                "report bug, abi vec does not have {}",
+                COMMIT_PARENT_FINALITY_FUNC_NAME
+            )
+        })?;
+
+    let validators = validator_set.validators.unwrap_or_default();
+
+    let mut addresses = vec![];
+    let mut weights = vec![];
+    for validator in validators {
+        let raw_address = validator.worker_addr.unwrap_or(validator.addr);
+        let addr = Address::from_str(&raw_address)?;
+        addresses.push(convert_addr(addr));
+        weights.push(U256::from_dec_str(&validator.weight)?);
+    }
+
+    let data = ethers::contract::encode_function_data(
+        commit_function,
+        gateway_router_facet::CommitParentFinalityCall {
+            finality: gateway_router_facet::ParentFinality::try_from(finality)?,
+            validators: addresses,
+            weights,
+        },
+    )?;
+
+    Ok(data.to_vec())
+}
+
+pub fn encode_get_latest_parent_finality() -> anyhow::Result<Vec<u8>> {
+    let function = gateway_getter_facet::GATEWAYGETTERFACET_ABI
+        .functions
+        .get(GET_LATEST_PARENT_FINALITY_FUNC_NAME)
+        .ok_or_else(|| {
+            anyhow!(
+                "report bug, abi function map does not have {}",
+                GET_LATEST_PARENT_FINALITY_FUNC_NAME
+            )
+        })?
+        .get(0)
+        .ok_or_else(|| {
+            anyhow!(
+                "report bug, abi vec does not have {}",
+                GET_LATEST_PARENT_FINALITY_FUNC_NAME
+            )
+        })?;
+
+    let data = ethers::contract::encode_function_data(function, ())?;
+
+    Ok(data.to_vec())
+}
+
+pub fn decode_parent_finality_return(bytes: &[u8]) -> anyhow::Result<IPCParentFinality> {
+    let function = gateway_getter_facet::GATEWAYGETTERFACET_ABI
+        .functions
+        .get(GET_LATEST_PARENT_FINALITY_FUNC_NAME)
+        .ok_or_else(|| {
+            anyhow!(
+                "report bug, abi function map does not have {}",
+                GET_LATEST_PARENT_FINALITY_FUNC_NAME
+            )
+        })?
+        .get(0)
+        .ok_or_else(|| {
+            anyhow!(
+                "report bug, abi vec does not have {}",
+                GET_LATEST_PARENT_FINALITY_FUNC_NAME
+            )
+        })?;
```

Comment:
This is the third time these 15 lines appear in this module. It seems easy to turn it into a method to be called with an `Abi` and function name.